File: src/DataBase/CompositeFieldListPolicy.cc
//---------------------------------Spheral++----------------------------------//
// CompositeFieldListPolicy -- An implementation of UpdatePolicyBase which
// consists of a collection of individual Field policies that should match
// the Fields in a FieldList.
//
// Created by JMO, Sun Nov  3 14:11:32 PST 2013
//----------------------------------------------------------------------------//
#include "CompositeFieldListPolicy.hh"
#include "FieldListUpdatePolicyBase.hh"
#include "State.hh"
#include "StateDerivatives.hh"
#include "Field/FieldList.hh"
#include "Utilities/DBC.hh"

namespace Spheral {

using FieldSpace::FieldList;

//------------------------------------------------------------------------------
// Constructors.
//------------------------------------------------------------------------------
template<typename Dimension, typename ValueType>
CompositeFieldListPolicy<Dimension, ValueType>::
CompositeFieldListPolicy():
  FieldListUpdatePolicyBase<Dimension, ValueType>(),
  mPolicyPtrs() {
}

//------------------------------------------------------------------------------
// Destructor.
//------------------------------------------------------------------------------
template<typename Dimension, typename ValueType>
CompositeFieldListPolicy<Dimension, ValueType>::
~CompositeFieldListPolicy() {
}

//------------------------------------------------------------------------------
// Update the FieldList.
//------------------------------------------------------------------------------
template<typename Dimension, typename ValueType>
void
CompositeFieldListPolicy<Dimension, ValueType>::
update(const KeyType& key,
       State<Dimension>& state,
       StateDerivatives<Dimension>& derivs,
       const double multiplier,
       const double t,
       const double dt) {

  // Get the field name portion of the key.
  KeyType fieldKey, nodeListKey;
  StateBase<Dimension>::splitFieldKey(key, fieldKey, nodeListKey);
  REQUIRE(nodeListKey == UpdatePolicyBase<Dimension>::wildcard());

  // Find the FieldList for this key.
  FieldList<Dimension, ValueType> f = state.fields(fieldKey, ValueType());

  // Walk the paired Fields and policies.
  const unsigned numFields = f.numFields();
  CHECK(mPolicyPtrs.size() == numFields);
  for (unsigned i = 0; i != numFields; ++i) {
    KeyType fkey = StateBase<Dimension>::buildFieldKey(fieldKey, f[i]->nodeListPtr()->name());
    mPolicyPtrs[i]->update(fkey, state, derivs, multiplier, t, dt);
  }
}

//------------------------------------------------------------------------------
// Equivalence operator.
//------------------------------------------------------------------------------
template<typename Dimension, typename ValueType>
bool
CompositeFieldListPolicy<Dimension, ValueType>::
operator==(const UpdatePolicyBase<Dimension>& rhs) const {

  // We're only equal if the other guy is also an increment operator.
  const CompositeFieldListPolicy<Dimension, ValueType>* rhsPtr = dynamic_cast<const CompositeFieldListPolicy<Dimension, ValueType>*>(&rhs);
  if (rhsPtr == 0) return false;
  return (this->mPolicyPtrs == rhsPtr->mPolicyPtrs);
}

//------------------------------------------------------------------------------
// Add a new UpdatePolicy to this thing.
//------------------------------------------------------------------------------
template<typename Dimension, typename ValueType>
void
CompositeFieldListPolicy<Dimension, ValueType>::
push_back(const typename CompositeFieldListPolicy<Dimension, ValueType>::PolicyPointer policyPtr) {
  mPolicyPtrs.push_back(policyPtr);
}

}

File: src/PYB11/DataBase/CompositeFieldListPolicy.py
from PYB11Generator import *
from FieldListUpdatePolicyBase import *

@PYB11module("DataBase")
@PYB11holder("std::shared_ptr")
@PYB11template("Dimension", "ValueType")
class CompositeFieldListPolicy(FieldListUpdatePolicyBase):
    """CompositeFieldListPolicy -- An implementation of UpdatePolicyBase which
consists of a collection of individual Field policies that should match
the Fields in a FieldList."""

    PYB11typedefs = """
    using KeyType = typename IncrementFieldList<%(Dimension)s, %(ValueType)s>::KeyType;
"""

    #...........................................................................
    # Constructors
    def pyinit0(self):
        return

    #...........................................................................
    # Methods
    @PYB11virtual
    def update(self,
               key = "const KeyType&",
               state = "State<%(Dimension)s>&",
               derivs = "StateDerivatives<%(Dimension)s>&",
               multiplier = "const double",
               t = "const double",
               dt = "const double"):
        "Update a FieldList assoicated with the given key"
        return "void"

    @PYB11virtual
    def updateAsIncrement(self,
                          key = "const KeyType&",
                          state = "State<%(Dimension)s>&",
                          derivs = "StateDerivatives<%(Dimension)s>&",
                          multiplier = "const double",
                          t = "const double",
                          dt = "const double"):
        """An alternate method to be called when you want to specify that the derivative information
should be assumed to not necessarily be properly time-centered, and therefore you should 
only use time advancement ideas, no "replace" or more sophisticated approaches.
Default to just calling the generic method."""
        return "void"

    def push_back(self,
                  policy = "UpdatePolicyBase<%(Dimension)s>*"):
        "Add a new UpdatePolicy"
        return "void"
